src/Typechecker.hs

```diff
@@ -4,49 +4,233 @@ module Typechecker where
+import Data.Functor.Identity
 
 -- Type environment
-data Gamma = Gamma [(Identifier, Type)] deriving (Show)
+data Gamma = Gamma (Env, TldMap, TcDef, TcImp) deriving (Show)
+data Env = Env [(Identifier, Type)] deriving (Show)
+data TldMap = TldMap [(Type, [CDef])] deriving (Show)
+data TcDef = TcDef [(Identifier, [SignatureDef])] deriving (Show)
+data TcImp = TcImp [(Identifier, [SignatureImp])] deriving (Show)
 
-addEntry :: Identifier -> Type -> Gamma -> Gamma
-addEntry n t (Gamma l) = Gamma (l ++ [(n,t)])
+
+addEntryToEnv :: Identifier -> Type -> Gamma -> Gamma
+addEntryToEnv n t (Gamma (Env l, TldMap m, TcDef td, TcImp ti)) = Gamma (Env (l ++ [(n,t)]), TldMap m, TcDef td, TcImp ti)
+
+addTcDefToGamma :: Identifier -> [SignatureDef] -> Gamma -> Gamma
+addTcDefToGamma n s (Gamma (Env l, TldMap m, TcDef td, TcImp ti)) = Gamma (Env l, TldMap m, TcDef (td ++ [(n, s)]), TcImp ti)
+
+addTcImpToGamma :: Identifier -> [SignatureImp] -> Gamma -> Gamma
+addTcImpToGamma n s (Gamma (Env l, TldMap m, TcDef td, TcImp ti)) = Gamma (Env l, TldMap m, TcDef td, TcImp (ti ++ [(n, s)]))
+
+insertTcDefsToGamma :: Identifier -> [SignatureDef] -> Gamma -> Maybe Gamma
+insertTcDefsToGamma n s (Gamma (Env l, TldMap m, TcDef td, TcImp ti)) = do
+    case td == [] of
+        True -> return (Gamma (Env l, TldMap m, TcDef [(n, s)], TcImp ti))
+        False -> do
+            index <- findIndex (==n) [name | (name, sigs) <- td]
+            let newTcDef = (n, ([sigs | (name, sigs) <- td] !! index ++ s))
+            let (x,_:ys) = splitAt index td
+            let newTd = x ++ newTcDef : ys
+            return (Gamma (Env l, TldMap m, TcDef newTd, TcImp ti))
+
+insertTcImpsToGamma :: Identifier -> [SignatureImp] -> Gamma -> Maybe Gamma
+insertTcImpsToGamma n s (Gamma (Env l, TldMap m, TcDef td, TcImp ti)) = do
+    case ti == [] of
+        True -> return (Gamma (Env l, TldMap m, TcDef td, TcImp [(n, s)]))
+        False -> do
+            index <- findIndex (==n) [name | (name, imps) <- ti]
+            let newTcImps = (n, ([imps | (name, imps) <- ti] !! index ++ s))
+            let (x,_:ys) = splitAt index ti
+            let newTi = x ++ newTcImps : ys
+            return (Gamma (Env l, TldMap m, TcDef td, TcImp newTi))
 
 getType :: Identifier -> Gamma -> Maybe Type
-getType x (Gamma l) = do
+getType x (Gamma (Env l, _, _, _)) = do
     t <- lookup x gMap
     return t
     where
         gMap = fromList l
 
+getImpType :: Identifier -> Identifier -> Type -> Gamma -> Maybe (Maybe Type)
+getImpType tcn tcf i (Gamma (_, _, _, TcImp ti)) = do
+    let gMap = fromList ti
+    let tcImps = lookup tcn gMap
+    case tcImps of
+        Nothing -> return Nothing
+        Just imps -> do
+            case lookup (tcf, i) gMap2 of
+                Nothing -> return Nothing
+                Just foundOut -> return (Just foundOut)
+            where
+                gMap2 = fromList (map getSigImpNameInOut imps)
+
+
+    -- case lookup tcn gMap of
```

Comment:
Should this be commented out? If its not needed it should be deleted

Reply:
Thanks, fixed